### DISCUSSION TITLE
How to connect to AWS IoT with custom authorizer?

### DISCUSSION THREAD

**Discussion by Karl-Styner** [emqx/MQTTX#1630] [Q&A]
I'm wasting my day trying to connect to AWS IoT Core. Please help me.

**Comment by ysfscream** [comment]
Hi, but could you please provide a specific description of your issue first?

**Reply by Karl-Styner** [reply]
I'm testing a custom authorizer according to the content of 
https://docs.aws.amazon.com/iot/latest/developerguide/custom-auth.html#custom-auth-mqtt

According to this content, the x-amz-customauthorizer-name should be passed correctly for the custom authorizer registered Lambda function to be invoked. However, it seems that this value is not being passed correctly no matter where I put it.
I'm using the latest MQTTX for macOS and have tried all possible combinations, but I have not been able to successfully connect in any way.

**Reply by ysfscream** [reply]
To successfully connect to AWS IoT Core using a custom authorizer with MQTTX, you need to ensure the following:

1. **Port and Protocol**: Ensure your device connects through port 443 and utilizes the Application Layer Protocol Negotiation (ALPN) TLS extension with a value of 'mqtt', and the Server Name Indication (SNI) extension with the hostname of your AWS IoT Core data endpoint.

2. **URL Encoding**: Parameters such as `x-amz-customauthorizer-name`, `x-amz-customauthorizer-signature`, and `token-key-name` should be URL encoded to prevent transmission errors.

3. **Base64 Encoding**: The data sent in the password field should be base64 encoded, which your Lambda function will then decode.

If you've followed all these settings and are still unable to connect, consider checking the following:

- Ensure that your custom authorizer is correctly configured in AWS Lambda.
- Verify that your token and signature are correctly generated and conform to expectations.
- Check the format of your query strings and other parameters in MQTTX to ensure they are exactly as required.

Here is an example format for passing the necessary parameters in the MQTT username field:
```
username?x-amz-customauthorizer-name=authorizer-name&x-amz-customauthorizer-signature=token-signature&token-key-name=token-value
```

<img width="1137" alt="image" src="https://github.com/emqx/MQTTX/assets/21299158/ea98f405-9475-47d7-9ac6-bd67dba295ae">
